numba/tests/doc_examples/test_interval_example.py

```diff
@@ -0,0 +1,221 @@
+
+        @register_model(IntervalType)
+        class IntervalModel(models.StructModel):
+            def __init__(self, dmm, fe_type):
+                members = [('lo', types.float64),
+                           ('hi', types.float64),]
+                models.StructModel.__init__(self, dmm, fe_type, members)
+        # magictoken.interval_model.end
+
+        # magictoken.interval_attribute_wrapper.begin
+        from numba.extending import make_attribute_wrapper
+
+        make_attribute_wrapper(IntervalType, 'lo', 'lo')
+        make_attribute_wrapper(IntervalType, 'hi', 'hi')
+        # magictoken.interval_attribute_wrapper.end
+
+        # magictoken.interval_overload_attribute.begin
+        from numba.extending import overload_attribute
+
+        @overload_attribute(IntervalType, "width")
+        def get_width(interval):
+            def getter(interval):
+                return interval.hi - interval.lo
+            return getter
+        # magictoken.interval_overload_attribute.end
+
+        # magictoken.interval_lower_builtin.begin
+        from numba.extending import lower_builtin
+        from numba.core import cgutils
+
+        @lower_builtin(Interval, types.Float, types.Float)
+        def impl_interval(context, builder, sig, args):
+            typ = sig.return_type
+            lo, hi = args
+            interval = cgutils.create_struct_proxy(typ)(context, builder)
+            interval.lo = lo
+            interval.hi = hi
+            return interval._getvalue()
+        # magictoken.interval_lower_builtin.end
+
+        # magictoken.interval_unbox.begin
+        from numba.extending import unbox, NativeValue
+        from contextlib import contextmanager, ExitStack
+
+        @contextmanager
+        def early_exit_if(builder, stack: ExitStack, cond):
+            """
+            Emit code similar to::
+
+                if (cond) {
+                    <body>
+                    return;
+                }
+                <everything after this call>
+
+            However, this "return" will break out of the current `ExitStack`
+            rather than out of the whole function. This gives us a native
+            functionality similar to that of a try-except block.
+            """
+            then, otherwise = stack.enter_context(
+                builder.if_else(cond, likely=False)
+            )
+            with then:
+                yield
+            stack.enter_context(otherwise)
+
+        def early_exit_if_null(builder, stack, obj):
+            return early_exit_if(builder, stack, cgutils.is_null(builder, obj))
+
+        @unbox(IntervalType)
+        def unbox_interval(typ, obj, c):
+            """
+            Convert a Interval object to a native interval structure.
+            """
+            is_error_ptr = cgutils.alloca_once_value(c.builder,
+                                                     cgutils.false_bit)
+
+            with ExitStack() as stack:
```

Comment:
I remember @stuartarchibald suggesting we could have a generalized solution for such Exit-if-null strategy. The main issue with this particular approach is that we can't abstract out the context definitions and exits. 

Is there anything that comes to mind regarding how we could do this with a more simpler method ?